content/courses/advanced-ada/chapters/aggregates.rst

```diff
@@ -435,14 +435,339 @@ consider using interfaces instead of null records. We'll discuss this topic
+specified a default value for component :ada:`Z` in the declaration of the
+:ada:`Point_3D` type, so we cannot rely on any specific value being assigned to
+that component when using :ada:`others => <>`.
+
+:ada:`with null record`
+~~~~~~~~~~~~~~~~~~~~~~~
+
+We can also use extension aggregates with null records. Let's focus on the
+:ada:`P_3D_Ext` object of :ada:`Point_3D_Ext` type. This object is declared in
+the :ada:`Show_Points` procedure of the next code example.
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Aggregates.Extension_Aggregate_Points
+
+    package Points.Extensions is
+
+       type Point_3D_Ext is new Point_3D with null record;
+
+    end Points.Extensions;
+
+    with Points;            use Points;
+    with Points.Extensions; use Points.Extensions;
+
+    procedure Show_Points is
+       P_3D     : Point_3D;
+       P_3D_Ext : Point_3D_Ext;
+    begin
+       P_3D := (X => 0.0, Y => 0.5, Z => 0.4);
+
+       P_3D_Ext := (P_3D with null record);
+       Display (P_3D_Ext);
+    end Show_Points;
+
+The :ada:`P_3D_Ext` object is of :ada:`Point_3D_Ext` type, which is declared in
+the :ada:`Points.Extensions` package and derived from the :ada:`Point_3D` type.
+Note that we're not extending :ada:`Point_3D_Ext` with new components, but
+using a null record instead in the declaration. Therefore, as the
+:ada:`Point_3D_Ext` type doesn't own any new components, we just write
+:ada:`(P_3D with null record)` to initialize the :ada:`P_3D_Ext` object.
+
+Extension aggregates and descendent types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the examples above, we've been initializing objects of descendent types by
+using objects of ascending types in extension aggregates. We could, however, do
+the opposite and initialize objects of ascending types using objects of
+descendent type in extension aggregates. Consider this code example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Aggregates.Extension_Aggregate_Points
+
+    with Points; use Points;
+
+    procedure Show_Points is
+       P_2D : Point_2D;
+       P_3D : Point_3D;
+    begin
+       P_3D := (X => 0.5, Y => 0.7, Z => 0.3);
+       Display (P_3D);
+
+       P_2D := (Point_1D (P_3D) with Y => 0.3);
+       Display (P_2D);
+    end Show_Points;
+
+Here, we're using :ada:`Point_1D (P_3D)` to select the :ada:`Point_1D` view of
+an object of :ada:`Point_3D` type. At this point, we have specified the
+:ada:`Point_1D` part of the aggregate, so we still have to specify the
+remaining components of the :ada:`Point_2D` type |mdash| the :ada:`Y`
+component, to be more specific. When we do that, we get the appropriate
+aggregate for the :ada:`Point_2D` type. In summary, by carefully selecting the
+appropriate view, we're able to initialize an object of ascending type
+(:ada:`Point_2D`), which contains less components, using an object of a
+descendent type (:ada:`Point_3D`), which contains more components.
 
```

Comment:
Really nice use of the view concept throughout!  Good work!

Reply:
Thanks! :-)